docs/TICCL-lexstat.1
.TH TICCL-lexstat 1 "2020 mar 09"

.SH NAME
TICCL-lexstat \- an alphabet file and character confusion file generator

.SH SYNOPSIS

TICCL-lexstat [options] FILE

.SH DESCRIPTION
.B TICCL-lexstat
Step 1: reads a language specific lexicon and creates an alphabet file based on character frequencies. Step 2: reads an alphabet file ( previously generated by
.B TICCL-lexstat
) and creates first a file with a frequency list of all character
values, then - after 'clipping' beneath a particular character frequency: a list of anagram hash values for all character confusion pairs up to the specified 'LD'

Character frequencies lower than 'clip' (option \-\-clip) are then skipped,

.SH OPTIONS
.B \-\-clip
clip
.RS
skip characters less frequent than 'clip'
.RE
.B \-\-LD
depth
.RS
Generate confusions up to a Levenshtein distance of 'depth' (default 2)

When LD=0, only a character frequency file is produced. This is the necessary first pass.
.RE

.B \-\-all
.RS
Output all different character confusion sequences for each character confusion anagram value. Default only the first one is shown.
.RE

.B \-\-diac
.RS
Output an extra file with confusions for diacritics
shown.
.RE

.B \-\-separator
sep
.RS
add 'sep' as a character to the character list. This is required for word bigrams or trigrams and replaces the 'space' character.
.RE

.B \-o
outfile
.RS
Base the output file name(s) on 'outfile'. Default the name of the inputfile is used. Use this also when the directory of the inputfile is read\-only
.RE

.B \-v
or
.B \-\-verbose
.RS
display extra information about the run
.RE

.B \-h
or
.B \-\-help
.RS
display usage information
.RE

.B \-V
.RS
Show VERSION
.RE


.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot lamasoftware@science.ru.nl

Martin Reynaert reynaert@uvt.nl


NEWS
ticcltools 0.6 2018-06-05
[Ko vander Sloot]
Intermediate release, with a lot of new code to handle N-grams
Also a lot of refactoring is done, for more clear and maintainable code.
This is work in progress still.

* TICCL-unk:
  - more extensive acronym detection
  - fixed artifreq problems in 'clean' punctuated words
  - added filters for 'unwanted' characters
  - added a ligature filter to convert evil ligatures
  - normalize all hyphens to a 'normal' one (-)
  - use a better definition of punctuation (unicode character class is not
    good enough to decide)

* TICCL-lexstat:
  - the 'separator' symbol should get freq=0, so it isn't counted
  - the clip value is added to the output filename

* TICCL-indexer:
  - indexer and indexerNT now produce the same output, using different
    strategies when a --foci files is used.

* TICCL-LDcalc:
  major overhaul for n-grams
  - added a ngram point column to the output (so NOT backward compatible!)
  - produce a '.short' list for short word corrections
  - produce a '.ambi' file with a list of n-grams related to short words
  - prune a lot of ngrams from the output

* TICCL-rank:
 - output is sorted now
 - honor the ngram-points from the new LDcalc. (so NOT backward compatible!)

* TICCL-chain: new module to chain ranked files

* TICCL-lexclean:
  -added a -x option for 'inverse' alphabet

* TICCL-anahash:
  - added a --list option to produce a list of words and anagram values

[Maarten van Gompel]
* added metadata file: codemeta.json

ticcltools 0.5 2018-02-19
[Ko van der Sloot]
* updated configuration. also for Mac OSX
* use of more ticcutils stuff: diacritics filter
* added a TICCL-mergelex program
* the OMP_THREAD_LIMIT environment variable was ignored sometimes
* TICCL-unk:
   - fixed a problem in artifreq handling
   - changed acronym detection (work in  progress)
   - added -o option
  TICCL-lexstat:
   - added TTR output
   - added -o option
  TICCL-indexer
  - now also handles --foci file. with some speed-up
  - added a -t option
  TICCL-LDcalc:
  - be less picky on a few wrong lines in the data
* added some tests
* when libroaring is installed we built roaring versions of some modules (experimental)
* updated man pages

ticcltools 0.4 2017-04-04
[Ko van der Sloot]
* first official release.
  - added functions to test on Word2Vec datafiles
  - refactoring and modernizing stuff all around

ticcltools 0.3 2016-01-21
* upload ready
ticcltools 0.2 2016-01-14
[Ko van der Sloot]
* repository moved to GitHub
* Travis support
* first 'distributable' version
* added TICCL-stats program
* added W2V-near and W2V-dist programs bases on new libticcl lib


[0.1] Ko vd Sloot
 started autoconfiscating


src/TICCL-lexstat.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of ticcltools

  ticcltools is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  ticcltools is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/ticcltools/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/
#include <cstdlib>
#include <getopt.h>
#include <string>
#include <set>
#include <map>
#include <iostream>
#include <fstream>

#include "ticcutils/StringOps.h"
#include "ticcutils/CommandLine.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/Unicode.h"
#include "ticcl/ticcl_common.h"

#include "config.h"

bool verbose = false;

using namespace	std;
using namespace icu;

void create_output( string& name, const map<UChar,size_t>& chars,
		    string& orig, map<UnicodeString,bitType>& hashes,
		    int clip, UnicodeString& separator ){
  ofstream os( name );
  if ( !os ){
    cerr << "unable to open output file: " << name << endl;
    exit(EXIT_FAILURE);
  }
  multimap<size_t,UChar> reverse;
  map<UChar,size_t>::const_iterator it = chars.begin();
  bitType count = 0;
  bitType out_count = 0;
  while ( it != chars.end() ){
    if ( clip >= 0 && it->second < (size_t)clip ){
      out_count += it->second;
    }
    else {
      count += it->second;
      reverse.insert( make_pair(it->second,it->first) );
    }
    ++it;
  }
  os << "## Alphabetsize: " << reverse.size() + (separator.isEmpty()?0:1)
     << endl;
  os << "## Original file : " << orig << " with " << count
     << " accepted characters and " << out_count << " clipped characters."
     << endl;
  int start = 100;
  bitType hash = high_five( start );
  hashes.insert( make_pair( "*", hash ) );
  os << "# *\tdigits_and_punctuation\t" << hash << endl;
  start = 101;
  hash = high_five( start );
  hashes.insert( make_pair( "$", hash ) );
  os << "# $\tunknown_characters\t" << hash << endl;
  start = 102;
  int spec_cnt = 2;
  if ( !separator.isEmpty() ){
    ++spec_cnt;
    hash = high_five( start );
    hashes.insert( make_pair( separator, hash ) );
    os << separator << "\t0\t\t" << hash << endl;
    start = 103;
  }
  multimap<size_t,UChar>::const_reverse_iterator rit = reverse.rbegin();
  int out_cnt = 0;
  while ( rit != reverse.rend() ){
    hash = high_five( start );
    UnicodeString us( rit->second );
    hashes.insert( make_pair( us, hash ) );
    os << us << "\t" << rit->first << "\t" << hash << endl;
    ++out_cnt;
    ++start;
    ++rit;
  }
  cout << "created outputfile " << name
       << " with " << out_cnt << " hased lowercase characters, and " << spec_cnt
       << " special hashes for (*,$" << (separator.isEmpty()?"":","+separator)
       << ")" << endl;
}

void create_dia_file( const string& filename,
		      const map<UChar,size_t>& chars,
		      const map<UnicodeString,bitType>& hashes ){
  ofstream os( filename );
  map<UChar,size_t>::const_iterator it = chars.begin();
  while ( it != chars.end() ){
    UnicodeString us;
    us += it->first;
    UnicodeString ss = TiCC::filter_diacritics( us );
    if ( ss != us ){
      map<UnicodeString,bitType>::const_iterator hit = hashes.find( us );
      if ( hit == hashes.end() ){
	if ( verbose ){
	  cerr << "problem: " << us << " not in the hashes?" << endl;
	}
	++it;
	continue;
      }
      bitType h1 = hit->second;
      hit = hashes.find( ss );
      if ( hit == hashes.end() ){
	if ( verbose ){
	  cerr << "problem: " << ss << " not in the hashes?" << endl;
	}
	++it;
	continue;
      }
      bitType h2 = hit->second;
      bitType h;
      if ( h1 > h2 )
	h = h1 - h2;
      else
	h = h2 - h1;
      os << h << "#" << us << "~" << ss << endl;
    }
    ++it;
  }
  cout << "created a diacritic confusion file: " << filename << endl;
}

void meld_botsing( const multimap<bitType,UnicodeString>& mm, bitType h ){
  multimap<bitType,UnicodeString>::const_iterator it;
  map<set<UChar>,UnicodeString > ref;
  for ( it = mm.lower_bound(h); it != mm.upper_bound(h); ++it ){
    set<UChar> st;
    UnicodeString s = it->second;
    for( int i=0; i < s.length(); ++i ){
      st.insert( s[i] );
    }
    ref.insert( make_pair(st,s) );
  }
  cerr << "collision at hash " << h << ": ";
  map<set<UChar>,UnicodeString >::const_iterator sit=ref.begin();
  while ( sit != ref.end() ){
    cerr << sit->second << ";";
    ++sit;
  }
  cerr << endl;
}

void conditionally_insert( multimap<bitType,UnicodeString>& confusions,
			   bitType key,
			   const UnicodeString& value,
			   bool full ){
  if ( full ){
    confusions.insert( make_pair( key, value ) );
  }
  else if ( confusions.find( key ) == confusions.end() ){
    confusions.insert( make_pair( key, value ) );
  }
}

void generate_confusion( const string& name,
			 const map<UnicodeString,bitType>& hashes,
			 int depth,
			 bool full ){
  ofstream os( name );
  if ( !os ){
    cerr << "unable to open output file: " << name << endl;
    exit(EXIT_FAILURE);
  }
  multimap<bitType,UnicodeString> confusions;
  cout << "start : " << hashes.size() << " iterations " << endl;
  map<UnicodeString,bitType>::const_iterator it1 = hashes.begin();
  while ( it1 != hashes.end() ){
    cout << it1->first << " ";
    // deletions/inserts of 1
    UnicodeString s1 = it1->first + "~";
    conditionally_insert( confusions, it1->second, s1, full );
    map<UnicodeString,bitType>::const_iterator it2 = hashes.begin();
    while ( it2 != hashes.end() ){
      if ( it2 != it1 ){
	// 1-1 substitutions
	bitType div;
	if ( it2->second > it1->second ){
	  div = it2->second - it1->second;
	}
	else {
	  div = it1->second - it2->second;
	}
	UnicodeString s2 = it1->first + "~" + it2->first;
	conditionally_insert( confusions, div, s2, full );
      }
      ++it2;
    }
    if ( depth > 1 ){
      it2 = hashes.begin();
      while ( it2 != hashes.end() ){
	// 2-0 substitutions
	UnicodeString s20 = it1->first + it2->first + "~";
	bitType div = it2->second + it1->second;
	conditionally_insert( confusions, div, s20, full );
	map<UnicodeString,bitType>::const_iterator it3 = hashes.begin();
	while ( it3 != hashes.end() ){
	  if ( it3 != it2 && it3 != it1 ){
	    // 2-1 substitutions
	    UnicodeString s = it1->first + it2->first + "~" + it3->first;
	    bitType div;
	    bitType d1 = it1->second + it2->second;
	    bitType d2 = it3->second;
	    if ( d1 > d2 ){
	      div = d1 - d2;
	    }
	    else {
	      div = d2 - d1;
	    }
	    conditionally_insert( confusions, div, s, full );
	  }
	  if ( it2 != it1 && it3 != it1 ){
	    // 1-2 substitutions
	    UnicodeString s = it1->first + "~" + it2->first + it3->first;
	    bitType div;
	    bitType d1 = it1->second;
	    bitType d2 = it2->second + it3->second;
	    if ( d1 > d2 ){
	      div = d1 - d2;
	    }
	    else {
	      div = d2 - d1;
	    }
	    conditionally_insert( confusions, div, s, full );
	  }
	  map<UnicodeString,bitType>::const_iterator it4 = hashes.begin();
	  while ( it4 != hashes.end() ){
	    if ( it3 != it1 && it3 != it2 && it4 != it1 && it4 != it2 ){
	      // 2-2 substitutions
	      UnicodeString s = it1->first + it2->first + "~"
		+ it3->first + it4->first;
	      bitType div;
	      bitType d1 = it1->second + it2->second;
	      bitType d2 = it3->second + it4->second;
	      if ( d1 > d2 ){
		div = d1 - d2;
	      }
	      else {
		div = d2 - d1;
	      }
	      conditionally_insert( confusions, div, s, full );
	    }
	    if ( depth > 2 ){
	      // 3-0 substitutions
	      UnicodeString s30 = it1->first + it2->first + it3->first + "~";
	      bitType div = it1->second + it2->second + it3->second;
	      conditionally_insert( confusions, div, s30, full );
	      if ( it4 != it1 && it4 != it2 && it4 != it3 ){
		// 3-1 substitutions
		UnicodeString s = it1->first + it2->first + it3->first + "~"
		  + it4->first;
		bitType div;
		bitType d1 = it4->second;
		bitType d2 = it3->second + it2->second + it1->second;
		if ( d1 > d2 ){
		  div = d1 - d2;
		}
		else {
		  div = d2 - d1;
		}
		conditionally_insert( confusions, div, s, full );
	      }
	      if ( it2 != it1 && it3 != it1 && it4 != it1 ){
		// 1-3 substitutions
		UnicodeString s = it1->first + "~"
		  + it2->first + it3->first + it4->first;
		bitType div;
		bitType d1 = it1->second;
		bitType d2 = it2->second + it3->second + it4->second;
		if ( d1 > d2 ){
		  div = d1 - d2;
		}
		else {
		  div = d2 - d1;
		}
		conditionally_insert( confusions, div, s, full );
	      }
	      map<UnicodeString,bitType>::const_iterator it5 = hashes.begin();
	      while ( it5 != hashes.end() ){
		if ( it4 != it1 && it4 != it2 && it4 != it3 &&
		     it5 != it1 && it5 != it2 && it5 != it3 ){
		  // 3-2 substitutions
		  UnicodeString s = it1->first + it2->first + it3->first + "~"
		    + it4->first + it5->first;
		  bitType div;
		  bitType d1 = it5->second + it4->second;
		  bitType d2 = it3->second + it2->second + it1->second;
		  if ( d1 > d2 ){
		    div = d1 - d2;
		  }
		  else {
		    div = d2 - d1;
		  }
		  conditionally_insert( confusions, div, s, full );
		}
		if ( it3 != it1 && it3 != it2 &&
		     it4 != it1 && it4 != it2 &&
		     it5 != it1 && it5 != it2 ){
		  // 2-3 substitutions
		  UnicodeString s = it1->first + it2->first + "~" + it3->first
		    +it4->first + it5->first;
		  bitType div;
		  bitType d1 = it5->second + it4->second + it3->second;
		  bitType d2 = it2->second + it1->second;
		  if ( d1 > d2 ){
		    div = d1 - d2;
		  }
		  else {
		    div = d2 - d1;
		  }
		  conditionally_insert( confusions, div, s, full );
		}
		map<UnicodeString,bitType>::const_iterator it6 = hashes.begin();
		while ( it6 != hashes.end() ){
		  if ( it6 != it1 && it6 != it2 && it6 != it3 &&
		       it5 != it1 && it5 != it2 && it5 != it3 &&
		       it4 != it1 && it4 != it2 && it4 != it3 ){
		    // 3-3 substitutions
		    UnicodeString s = it1->first + it2->first + it3->first + "~"
		      +it4->first + it5->first + it6->first;
		    bitType d1 = it6->second + it5->second + it4->second;
		    bitType d2 = it3->second + it2->second + it1->second;
		    bitType div;
		    if ( d1 > d2 ){
		      div = d1 - d2;
		    }
		    else {
		      div = d2 - d1;
		    }
		    conditionally_insert( confusions, div, s, full );
		  }
		  ++it6;
		}
		++it5;
	      }
	    }
	    ++it4;
	  }
	  ++it3;
	}
	++it2;
      }
    }
    ++it1;
  }
  multimap<bitType,UnicodeString>::const_iterator cit = confusions.begin();
  if ( full ){
    bitType start=0;
    set<UnicodeString> unique;
    while ( cit != confusions.end() ){
      if ( cit->first != start ){
	// a new KWC starts
	if ( !unique.empty() ){
	  if ( unique.size() > 8 ){
	    meld_botsing( confusions, start );
	  }
	  os << start;
	  for ( const auto& un : unique ){
	    os << "#" << un;
	  }
	  os << endl;
	  unique.clear();
	}
	start = cit->first;
      }
      unique.insert( cit->second );
      ++cit;
    }
  }
  else {
    while ( cit != confusions.end() ){
      os << cit->first << "#" << cit->second << endl;
      ++cit;
    }
  }
  cout << endl << "generated confusion file " << name << endl;
}

void usage( const string& name ){
  cerr << "Usage:\t" << name << " [options] dictionary" << endl;
  cerr << "\t" << name << " will create a lowercased character frequency" << endl
       << "\t\tlist from a dictionary file," << endl
       << "\t\tand a character confusion file, based on that list." << endl;
  cerr << "\t-o 'name'\t create outputfile(s) with prefix 'name'" << endl;
  cerr << "\t--diac\t produces an extra diacritics confusion file (extension .diac)" << endl;
  cerr << "\t--clip\t 'clip' truncates the character file at frequency 'clip'" << endl;
  cerr << "\t--LD depth\t 1, 2 or 3. (default 2) The characterlength of the confusions." << endl;
  cerr << "\t\tWhen LD=0 only a frequency list is generated." << endl;
  cerr << "\t--separator=<sep> Add the 'sep' symbol to the alphabet." << endl;
  cerr << "\t--all\tfull output. Show ALL variants in the confusions file." << endl;
  cerr << "\t\tNormally only the first is shown." << endl;
  cerr << "\t-h or --help\t this message " << endl;
  cerr << "\t-v or --verbose\t give more details during run." << endl;
  cerr << "\t-V or --version\t show version " << endl;
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts;
  try {
    opts.set_short_options( "vVho:" );
    opts.set_long_options( "LD:,clip:,diac,all,separator:,help,verbose,version" );
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage( argv[0] );
    exit( EXIT_FAILURE );
  }
  string progname = opts.prog_name();
  if ( opts.extract('h' ) || opts.extract("help") ){
    usage( progname );
    exit(EXIT_SUCCESS);
  }
  verbose = opts.extract('v') || opts.extract("verbose");
  if ( opts.extract('V' ) || opts.extract("version") ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( argc < 2	){
    usage( progname );
    exit(EXIT_FAILURE);
  }

  string output_name;
  opts.extract( 'o', output_name );
  int depth = 2;
  string depthS = "2";
  int clip = -1;
  string clipS = "0";
  string diafile;
  bool full = opts.extract( "all" );
  bool stripdia = opts.extract( "diac" );
  string value;
  if ( opts.extract( "clip", clipS ) ){
    if ( !TiCC::stringTo(clipS,clip) ) {
      cerr << "illegal value for --clip (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
  }

  if ( opts.extract( "LD", depthS ) ){
    if ( !TiCC::stringTo(depthS,depth) ) {
      cerr << "illegal value for --LD (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    if ( depth <0 || depth > 3 ){
      cerr << "Invalid depth --LD " << depth << endl;
      exit(EXIT_FAILURE);
    }
  }
  UnicodeString separator;
  if ( opts.extract( "separator", separator ) ){
    if ( separator.length() != 1 ){
      cerr << "invalid separator, should be 1 unicode point!" << endl;
      exit( EXIT_FAILURE );
    }
  }

  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage(progname);
    exit(EXIT_FAILURE);
  }

  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.size() == 0 ) {
    cerr << "missing a dictionary inputfile!" << endl;
    exit( EXIT_FAILURE );
  }
  if ( fileNames.size() > 1 ) {
    cerr << "only one inputfile my be provided!" << endl;
    exit( EXIT_FAILURE );
  }
  string file_name = fileNames[0];
  ifstream is( file_name );
  if ( !is ){
    cerr << "unable to open dictionary file: " << file_name << endl;
    exit(EXIT_FAILURE);
  }
  string orig = TiCC::basename( file_name );
  if ( output_name.empty() ){
    output_name = file_name;
  }
  string lc_file_name = output_name +  ".clip" + clipS + ".lc.chars";
  if ( !TiCC::createPath( lc_file_name ) ){
    cerr << "unable to acces the path for '" << lc_file_name << "'" << endl;
    return EXIT_FAILURE;
  }
  string confusion_file_name = output_name + ".clip" + clipS + ".ld" + depthS + ".charconfus";
  if ( stripdia ){
    diafile = output_name + ".lc.diac";
  }

  map<UChar,size_t> lchars;
  string line;
  while ( getline( is, line ) ){
    UnicodeString us = TiCC::UnicodeFromUTF8( line );
    us.toLower();
    for ( int i = 0; i < us.length(); ++i ){
      ++lchars[us[i]];
    }
  }
  cout << "done reading" << endl;
  map<UnicodeString,bitType> hashes;
  create_output( lc_file_name, lchars, orig, hashes, clip, separator );
  if ( stripdia ){
    create_dia_file( diafile, lchars, hashes );
  }
  if ( depth > 0 ){
    generate_confusion( confusion_file_name, hashes, depth, full );
  }
  cout << "done!" << endl;
}
